Feature 52 - searching tags

jerushachua opened the pull request:
This is a feature implementation for issue #52 . 

I added a function which pulls the input from the search bar and displays only the matching input. The input is case sensitive, since Linux file systems are also case sensitive, but this can be easily changed if needed. 

Please feel free to help test any edge cases I may have missed.

sterlinglaird commented on app/misc/git.ts:
```diff
@@ -495,6 +495,56 @@ function createBranch() {
+    });
+}
+
+function displayTag(name, id, onclick) {
```
This function already exists in repo.ts?

jerushachua replied:
Yes - I thought repo.ts is only called when the repo is initialized. If this isn't the case, then I can just remove the duplicate function.

sterlinglaird commented on app/misc/git.ts:
```diff
@@ -495,6 +495,56 @@ function createBranch() {
+          return repo.getReferences(Git.Reference.TYPE.LISTALL);
+        }).then(function (refList) {
+          refList.sort();
+          bname = {};
```
Why are these objects getting cleared?

sterlinglaird replied:
And couldnt you just get the tag name from the tags object? that way we dont need to continue querying node git

jerushachua replied:
I wasn't sure if the branches and tags would be reloaded in real time. This only queries nodegit if the tags are being searched for. It's just to ensure we have the latest copy of tags/branches. But you're right that it would be more efficient to use the tags object, if it's up to date. I can test it out with the tags object and if it's working well, I submit a changeset.

sterlinglaird replied:
It probably doesn't matter too much, the object should be up to date though. Regardless, they shouldn't be cleared from within this function, thats probably more important.

sterlinglaird commented:
This isn't a requirement for completion of this issue but it wouldn't be very difficult to do:
The search feature would probably be a lot more usable if every time the search bar had input all the tags which have a matching prefix show up. That way if you type in "v1." you'd see "v1.0, v1.0.1, v1.2, v1.3 etc". The "ok" button could also get removed since that is only for creating a branch, which James doesn't actually want to happen through that menu anyways.

sterlinglaird commented on app/misc/git.ts:
```diff
@@ -495,6 +495,56 @@ function createBranch() {
+          refList.sort();
+          bname = {};
+          tags = {};
+          for (let i = 0; i < refList.length; i++) {
```
Theres a missing closing curly bracket for this for loop. My typescript compiler didn't like it

jerushachua replied:
should be fixed now!

SiRumCz commented:
Based on my experience on old(like ancient) AnuglarJS, I think they have built-in module or mechanics to be bable to filter the tag or branch name by typing partial names. Hope it could help you.

jerushachua commented:
Fixed all the issues that were mentioned. I removed the OK button and searching is now responsive to user input. Let me know if there are any issues that I missed in my own testing.

colinwerner commented on app/components/header/header.component.ts:
```diff
@@ -1,98 +1,98 @@
-  }
-
-}
+import { Component } from "@angular/core";
```
What's with all the whitespace differences?

jerushachua replied:
I don't know why git detects this as a diff.

colinwerner replied:
Just revert it.

colinwerner commented on app/misc/git.ts:
```diff
@@ -495,6 +495,37 @@ function createBranch() {
+          for (let i = 0; i < refList.length; i++) {
+
+            // strip name for readability
+            let refName = refList[i].name().split("/")[refList[i].name().split("/").length - 1];
```
Logically you can move this into the if statement?

Also, can you state what format the name is, please?
ie
name is "/blah/blah/blah/name" and we want to extract "name"?

jerushachua replied:
refName is the readable, string format of the tag name. The string returned from nodegit needs to be stripped before it can be used. 

Which if statement are you referring to?

colinwerner replied:
It's an optimization. You are doing the parsing of the name BEFORE checking if it is a tag. You should check if it is a tag before parsing the name. Hence move let refName =... into the if (refList[i].isTag()) {

colinwerner commented on app/misc/git.ts:
```diff
@@ -495,6 +495,37 @@ function createBranch() {
+            let refName = refList[i].name().split("/")[refList[i].name().split("/").length - 1];
+
+            if (refList[i].isTag()){
+              if (refName.indexOf( document.getElementById("tag-name").value ) > -1) {
```
Just to be sure, if you find the refname in the tagname then it returns > -1?

jerushachua replied:
Yes, js has a built-in function which helps you compare strings. This will check if refName is a substring of the tag we're currently comparing to.

colinwerner commented on app/misc/git.ts:
```diff
@@ -495,6 +495,37 @@ function createBranch() {
+              if (refName.indexOf( document.getElementById("tag-name").value ) > -1) {
+                document.getElementById(refName).setAttribute("style","display:block");
+              } else {
+                document.getElementById(refName).setAttribute("style","display:none");
```
lines 515 and 517 could be refactored?
if > -1
  attr = "display:block"
else
  attr = "display:none"
document.getelementbyid(refname).setattribute("style", attr)

jerushachua replied:
Sure, done.

colinwerner commented on app/misc/git.ts:
```diff
@@ -810,12 +841,12 @@ function displayModifiedFiles() {
           if (document.getElementById("modified-files-message") !== null) {
             let filePanelMessage = document.getElementById("modified-files-message");
-            filePanelMessage.parentNode.removeChild(filePanelMessage);
+            filePanelMessage.parentNode.Child(filePanelMessage);
```
I'm just curious about this, what necessitates this change?

jerushachua replied:
I changed this to test during the functionality of removing nodes. There wasn't any effect by the change here, but I forgot to change it back. I'll change it right now and update the files.

colinwerner commented on app/misc/git.ts:
```diff
@@ -495,6 +495,37 @@ function createBranch() {
+          // grab the list of references - these could be branches or tags
+          return repo.getReferences(Git.Reference.TYPE.LISTALL);
+        }).then(function (refList) {
+          refList.sort();
```
Is the sort necessary if we're doing an indexed search on the name?

colinwerner commented on app/misc/git.ts:
```diff
@@ -495,6 +495,37 @@ function createBranch() {
+                document.getElementById(refName).setAttribute("style","display:none");
+              }
+            }
+            lastRefList = refList.slice(); // update lastRefList
```
Does slice just return a copy of the array? What exactly is lastRefList used for?

jerushachua replied:
This was used in the original cleaning of the array when we first grab the tags. I can remove it as it's not needed here.
